### ISSUE TITLE
GSON: NoSuchField exception for generated TypeAdapter, Types can't access private field via getField()

### ISSUE THREAD

**Issue by pogo19** [immutables/immutables#367]
I have a class Event defined as this:

```java
@Immutable
@Gson.TypeAdapters
public abstract class Event<K, T> {

	@Parameter
	public abstract String getUuid();

	@Parameter
	public abstract long getTimestamp();

	@Parameter
	public abstract int getSchemaVersion();

	@Parameter
	public abstract EventType getType();

	@Parameter
	public abstract String getOriginator();

	@Parameter
	public abstract Any<K> getPayloadId();

	@Parameter
	public abstract Any<T> getPayload();

}
```

It uses another class (not managed by Immutables):
```java
public class Any<T> {

	public static <T> Any<T> of(Type type, T value) {
		return new Any<>(type, value);
	}

	private final Type type;
	private final T value;

	private Any(Type type, T value) {
		this.type = type;
		this.value = value;
		System.out.println("!!!! New ANY object, type " + type.toString()
				+ ", value: " + value); // FIXME
	}

	public T get() {
		return value;
	}

	public Type getType() {
		return type;
	}

	@SuppressWarnings("unchecked")
	public Class<T> getRawType() {
		return (Class<T>) TypeToken.of(type).getRawType();
	}

	@Override
	public String toString() {
		return "Any{type=" + type + ", value=" + value + "}";
	}
}
```

When Immutables' GSON TypeAdapters are used I get his exception:
```java
java.lang.AssertionError: Cannot get generic type for EventTypeAdapter.payloadIdTypeSample
	at org.immutables.gson.adapter.Types.tokenFor(Types.java:39)
	at <my package>.entities.storage.GsonAdaptersEvent$EventTypeAdapter.<init>(GsonAdaptersEvent.java:48)
	at <my package>.entities.storage.GsonAdaptersEvent.create(GsonAdaptersEvent.java:26)
	at com.google.gson.Gson.getAdapter(Gson.java:407)
	at <my package>.services.data.impl.AbstractGsonJacksonSerializerServiceImpl.getTypeAdapter(AbstractGsonJacksonSerializerServiceImpl.java:93)
	at <my package>.services.data.impl.AbstractGsonJacksonSerializerServiceImpl.writeValue(AbstractGsonJacksonSerializerServiceImpl.java:84)
	at $SerializerService_87ed30ba5736.writeValue(Unknown Source)
	at $SerializerService_87ed30ba5734.writeValue(Unknown Source)
	at <my package>.services.data.impl.YAMLSerializerServiceTest.readWriteGenerics(YAMLSerializerServiceTest.java:58)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:86)
	at org.testng.internal.Invoker.invokeMethod(Invoker.java:643)
	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:820)
	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1128)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:129)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:112)
	at org.testng.TestRunner.privateRun(TestRunner.java:782)
	at org.testng.TestRunner.run(TestRunner.java:632)
	at org.testng.SuiteRunner.runTest(SuiteRunner.java:366)
	at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:361)
	at org.testng.SuiteRunner.privateRun(SuiteRunner.java:319)
	at org.testng.SuiteRunner.run(SuiteRunner.java:268)
	at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:52)
	at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:86)
	at org.testng.TestNG.runSuitesSequentially(TestNG.java:1244)
	at org.testng.TestNG.runSuitesLocally(TestNG.java:1169)
	at org.testng.TestNG.run(TestNG.java:1064)
	at org.testng.remote.AbstractRemoteTestNG.run(AbstractRemoteTestNG.java:126)
	at org.testng.remote.RemoteTestNG.initAndRun(RemoteTestNG.java:137)
	at org.testng.remote.RemoteTestNG.main(RemoteTestNG.java:58)
Caused by: java.lang.NoSuchFieldException: payloadIdTypeSample
	at java.lang.Class.getField(Class.java:1703)
	at org.immutables.gson.adapter.Types.tokenFor(Types.java:35)
	... 32 more
```

I tried debugging this problem and it looks that at `org.immutables.gson.adapter.Types.tokenFor(Types.java:35)` is used `Class.getField(..)` but that doesn't find generated `private final` fields. Using `getDeclaredField(..)` should solve the problem.

Tested with versions 2.2.2, 2.2.4, 2.2.7. Also using Guava 19 and JDK 1.8 source and binary compatibility.

**Comment by elucash** [comment]
It was designed to be public, just doesn't make sense for private as the class itself is private, but in the course of action it became private and there was not test case to catch this. Fixed in 2.2.8-SNAPSHOT, to be released in a day or two. Thank you for revealing this!

**Comment by pogo19** [comment]
Thanks!

**Comment by elucash** [comment]
The fix released in 2.2.8 Thank you!
